### PR TITLE
AC-6678: Allow a merge-to-deploy workflow to an AWS Stack e.g pre-staging

### PR DESCRIPTION
**Pull request by VincentHokie**
#### Changes introduced in [AC-6678](https://masschallenge.atlassian.net/browse/AC-6678)
- add logic that triggers an an ecs deploy when development is built

#### How to test
- rerunning the last [impact-api build](https://travis-ci.org/masschallenge/impact-api/builds/536757005?utm_source=github_status&utm_medium=notification) for this branch a deploy will be triggered at the end
- you can "see" the deployment in action on [ECS](https://console.aws.amazon.com/ecs/home?region=us-east-1#/clusters/pre-staging/services/testing-impact-service/tasks). you will notice that the re are now 2 tasks running showing that a deployment is in progress. The way deployments work, is a new task is created (to make a total of 2) and once that is healthy the old one is destroyed leaving the new task with the new code.
- also on the [events](https://console.aws.amazon.com/ecs/home?region=us-east-1#/clusters/pre-staging/services/testing-impact-service/events) section on ECS, you will notice a sequence of events that show a deployment took place. i.e.

```
service testing-impact-service has reached a steady state.
service testing-impact-service has stopped 1 running tasks: task a5d5b70b-f767-42e9-a444-97cdc367ad99.
service testing-impact-service has begun draining connections on 1 tasks.
service testing-impact-service deregistered 1 targets in target-group impact-pre-staging-target-group
service testing-impact-service registered 1 targets in target-group impact-pre-staging-target-group
service testing-impact-service has started 1 tasks: task 822f382f-f5b7-4606-8fc6-8ff4fb3c6eb3.
service testing-impact-service has reached a steady state.
```

ps, the logs are read from bottom to top

### CODE REVIEW

File: Makefile
Comment (kitaka) [inline Makefile]: @VincentHokie why do we need terraform? My understanding was that Terraform is used for infrastructure automation and with that in mind, this doesn't seem to be the right application.
Can't we trigger some sort of refresh with just the `aws cli`?

Comment (VincentHokie) [reply Makefile]: there's a cli trigger that is done [here](https://github.com/masschallenge/impact-api/pull/294/files#diff-b67911656ef5d18c4ae36cb6741b7965R487) incase the infrastructure changes themselves don't lead to a deploy.

if e.g. we edit our [task definition](https://github.com/masschallenge/infrastructure/blob/development/super_stack/ecs-cluster/impact/task-def.json) (with say a new secret), terraform would do 2 things.
- update the infrastructure &
- perform a deploy on ECS

Adding terraform here has the benefit of us being able to test continuous changes to our infrastructure as well.

Comment (kitaka) [reply Makefile]: In the spirit of separation of concerns. Would it not make more sense to have the terraform on the infrastructure repo and keeping this to the simple task of triggering a deploy? 
The simpler, the easier it will be to maintain by people without terraform experience.

Comment (VincentHokie) [reply Makefile]: sure, makes sense. done.

Comment (VincentHokie) [reply Makefile]: note though, that were going to have several stacks up built off e.g master. So a single travis build on the infrastructure repo updating > 1 AWS Stack won't be an easy thing to accomplish.

But then again, infra changes aren't regular. So we'll see how it goes.

Comment (kitaka) [reply Makefile]: Do we really need to have continuous deployment on infrastructure? A bug in Accelerate code might not be as disruptive as one in the Infrastructure code. I wonder what the best practices are in the infrastructure as code community.

Comment (VincentHokie) [reply Makefile]: I think we can hold off on CD for infra for now.

**Comment by kitaka** [comment]
@VincentHokie I think a way to test is necessary. Is it possible to have some sort of `ping repo` that works exactly like development - in that when something is merged to it, the deploy happens?

Also, I believe a new task should be created on ECS upon a successful deployment with a start date that should match the time of deployment.

**Comment by VincentHokie** [comment]
A task definition does not `need` to be created since nothing has changed. The one on staging/prod does this because it uses a different tool. and that tool did not correctly create the task definition on deployment, which is what led to terraform.

About the testing 😩 forgot to update this repo. The rest have this already. Updating now.

**Comment by VincentHokie** [comment]
Updated.

**Comment by pbx** [comment]
Tagged "Don't merge" based on this: "Before this PR is merged, I will need to stop AC-6678 from performing a continuous deploy situation"

**Comment by kitaka** [comment]
> Also, I believe a new task should be created on ECS upon a successful deployment with a start date that should match the time of deployment

I meant it as a way of verifying that the deployment works!

**Review by kitaka — APPROVED** [review]

**Comment by VincentHokie** [comment]
> I meant it as a way of verifying that the deployment works!

ah 🤦‍♂

File: Makefile
Comment (shanbady) [inline Makefile:475]: this may still be relevant but im not going to block on it

File: Makefile
Comment (shanbady) [inline Makefile:469]: is this hard-coded branch name needed once merged?

File: Makefile
Comment (shanbady) [inline Makefile:471]: could we/should we store this script on our own/this repo?

File: web/impact/impact/settings.py
Comment (shanbady) [inline web/impact/impact/settings.py]: does the "." just before "masschallenge" need to be escaped?

**Review by shanbady — CHANGES_REQUESTED** [review]

File: web/impact/impact/settings.py
Comment (VincentHokie) [reply web/impact/impact/settings.py]: pre-staging works without escaping the `.`

so I'd say, no. We don't need to escape this

File: Makefile
Comment (VincentHokie) [reply Makefile:469]: the hard-coded `AC-6678` can go..and there's a [ticket](https://masschallenge.atlassian.net/browse/AC-6754) to clean this up once merged.

File: Makefile
Comment (VincentHokie) [reply Makefile:471]: We can, I just think in case AWS changes its API for some reason..that overhead would fall on us to update. Pulling it from the remote allows us to always have a "valid" script

File: web/impact/impact/settings.py
Comment (shanbady) [reply web/impact/impact/settings.py]: while it will validate on the cases we care about without the escape this regex matches any domain so long as it ends in "masschallenge.org" (ex: https://maliciousdomainmasschallenge.org)

Comment (VincentHokie) [reply web/impact/impact/settings.py]: ohh, I see what you mean! good catch! fixing now

Comment (VincentHokie) [reply web/impact/impact/settings.py]: fixed.

**Review by shanbady — APPROVED** [review]

**Pull request merged**
